src/macros-by-example.adoc

```diff
@@ -0,0 +1,823 @@
+
+[{mrk}]
+----
+(*public* price
+  [a, c]                          // signature
+  { amount: a, currency: c })     // template
+----
+----
+(:price 99 USD) ⇒ { amount: 99, currency: USD }
+----
+
+This macro has a signature that declares two parameters, named `a` and `c`, and it
+therefore accepts two arguments when invoked.
+
+NOTE: We are careful to distinguish between the
+views from “inside” and “outside” the macro: _parameters_ are the names used by a macro’s
+implementation to refer to its expansion-time inputs, while _arguments_ are the data provided to a
+macro at the point of invocation. In other words, we have “formal” parameters and “actual”
+arguments.
+
+The struct
+in this macro is our first non-trivial _template_, an expression in Ion’s new domain-specific language
+for defining macro functions.  This expression language treats Ion scalar values (except for
+symbols) as literals, giving the decimal in ``pi``’s template its intended meaning.  Expressions
+that are structs are interpreted as _quasi-literals_: the field names are literal, but the field
+“values” are arbitrary expressions.  This is why the `amount` and `currency` field names show up
+as-is in the expansion.  The sub-expressions in this template demonstrate that the expression
+language treats symbols as variable references.  Here, `a` and `c` in the template refer to the
+parameters of the macro, and during expansion they are “filled in” with the values supplied by
+the invocation of the macro.
+
+The template language also treats lists quasi-literally, and every element inside the list is an
+expression.  Here’s a silly macro to illustrate:
+
+[{mrk}]
+----
+(*public* reverse [a, b] [b, a])
+----
+----
+(:reverse first {amount:a, currency:c}) ⇒ [{amount:a, currency:c}, first]
+----
+
+Note that the `a` in the E-expression is _not_ part of the
+expression language and not a reference to the macro’s first parameter or any other named entity.
+From the point of view of ``reverse``’s template, the input struct is literal data.
+
+E-expressions can nest, so we could also encode the same data using `price`:
+
+----
+(:reverse first (:price a c)) ⇒ [{amount:a, currency:c}, first]
+----
+
+
+=== Invoking Macros from Templates
+
+So the template language treats symbols as parameter references, other scalars as literals, and
+structs and lists as quasi-literals, but what about S-expressions?
+
+[{mrk}]
+----
+(*public* website_url
+  [path]
+  (make_string "https://www.amazon.com/" path))
+----
+
+Template expressions that are S-expressions are operator invocations, where the operators are
+either special forms or macro invocations.  In this case, it’s the latter: an invocation of the
+system macro (_i.e._, built-in function) `make_string`, which concatenates its arguments to produce
+a single string:
+
+----
+(:website_url "gp/cart") ⇒ "https://www.amazon.com/gp/cart"
+----
+
+In the template language, macro invocations can appear almost anywhere:
```

Comment:
What are the rules for macro resolution in the templates? Can I refer to a macro from another module? How so?

Reply:
Good question, it'll be in a forthcoming doc covering the module system.